File: src/main/java/cn/chahuyun/economy/plugin/FishManager.java
package cn.chahuyun.economy.plugin;

import cn.chahuyun.economy.HuYanEconomy;
import cn.chahuyun.economy.entity.fish.Fish;
import cn.chahuyun.hibernateplus.HibernateFactory;
import cn.hutool.poi.excel.ExcelReader;
import cn.hutool.poi.excel.ExcelUtil;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 鱼管理
 *
 * @author Moyuyanli
 * @date 2022/12/9 14:53
 */
public class FishManager {

    /**
     * 归类整理的鱼
     */
    private static final Map<Integer, List<Fish>> fishMap = new HashMap<>();

    private FishManager() {
    }

    /**
     * 初始化鱼管理
     */
    public static void init() {
        List<Fish> fishList = HibernateFactory.selectList(Fish.class);
        if (fishList == null || fishList.isEmpty()) {
            reloadFish();
            return;
        }
        readFish(fishList);
    }

    /**
     * 获取对应的鱼等级
     *
     * @param fishLevel 鱼的等级
     * @return 对应等级的鱼集合
     * @author Moyuyanli
     * @date 2022/12/9 15:49
     */
    public static List<Fish> getLevelFishList(int fishLevel) {
        return fishMap.get(fishLevel);
    }

    /**
     * 从resources里面读取鱼数据
     */
    private static void reloadFish() {
        HuYanEconomy instance = HuYanEconomy.INSTANCE;
        InputStream resourceAsStream = instance.getResourceAsStream("fish.xls");
        Map<String, String> map = new HashMap<>();
        map.put("等级", "level");
        map.put("名称", "name");
        map.put("描述", "description");
        map.put("单价", "price");
        map.put("最小尺寸", "dimensionsMin");
        map.put("最大尺寸", "dimensionsMax");
        map.put("尺寸1阶", "dimensions1");
        map.put("尺寸2阶", "dimensions2");
        map.put("尺寸3阶", "dimensions3");
        map.put("尺寸4阶", "dimensions4");
        map.put("难度", "difficulty");
        map.put("特殊标记", "special");
        ExcelReader reader = ExcelUtil.getReader(resourceAsStream);
        List<Fish> fishList = reader.setHeaderAlias(map).readAll(Fish.class);
        for (Fish fish : fishList) {
            HibernateFactory.merge(fish);
        }
        readFish(fishList);
    }


    private static void readFish(List<Fish> list) {
        for (Fish fish : list) {
            int level = fish.getLevel();
            if (fishMap.containsKey(level)) {
                fishMap.get(level).add(fish);
            } else {
                fishMap.put(level, new ArrayList<>() {{
                    add(fish);
                }});
            }
        }
    }

}

File: src/main/kotlin/cn/chahuyun/economy/plugin/FishManager.kt
package cn.chahuyun.economy.plugin

import cn.chahuyun.economy.HuYanEconomy
import cn.chahuyun.economy.data.proxy.EntityProxyRegistry
import cn.chahuyun.economy.model.fish.FishDto
import cn.chahuyun.economy.repository.FishRepository
import cn.hutool.poi.excel.ExcelReader
import cn.hutool.poi.excel.ExcelUtil
import java.io.InputStream

/**
 * 楸肩鐞? */
object FishManager {
    /**
     * 褰掔被鏁寸悊鐨勯奔
     */
    private val fishMap: MutableMap<Int, MutableList<FishDto>> = HashMap()

    @JvmStatic
    fun init() {
        val fishList = fishProxy.findAll()
        if (fishList.isEmpty()) {
            reloadFish()
            return
        }
        readFish(fishList)

        FishRepository.resetAllFishingStatus()
    }

    /**
     * 鑾峰彇瀵瑰簲鐨勯奔绛夌骇
     */
    @JvmStatic
    fun getLevelFishList(fishLevel: Int): List<FishDto> {
        return fishMap[fishLevel] ?: emptyList()
    }

    /**
     * 浠?resources 閲岄潰璇诲彇楸兼暟鎹?     */
    private fun reloadFish() {
        val instance = HuYanEconomy
        val resourceAsStream: InputStream = instance.getResourceAsStream("fish.xls") ?: return
        val map = hashMapOf(
            "绛夌骇" to "level",
            "鍚嶇О" to "name",
            "鎻忚堪" to "description",
            "鍗曚环" to "price",
            "最小尺寸" to "dimensionsMin",
            "最大尺寸" to "dimensionsMax",
            "尺寸1阈值" to "dimensions1",
            "尺寸2阈值" to "dimensions2",
            "尺寸3阈值" to "dimensions3",
            "尺寸4阈值" to "dimensions4",
            "闅惧害" to "difficulty",
            "鐗规畩鏍囪" to "special"
        )
        val reader: ExcelReader = ExcelUtil.getReader(resourceAsStream)
        val fishList: List<FishDto> = reader.setHeaderAlias(map).readAll(FishDto::class.java)
        fishProxy.saveAll(fishList)
        readFish(fishList)
    }

    private fun readFish(list: List<FishDto>) {
        fishMap.clear()
        for (fish in list) {
            val level = fish.level
            fishMap.getOrPut(level) { mutableListOf() }.add(fish)
        }
    }

    private val fishProxy
        get() = EntityProxyRegistry.get<FishDto>("fish") ?: error("鱼种代理器未初始化")
}
